homeassistant/components/iotty/cover.py

```diff
@@ -0,0 +1,218 @@
+            manufacturer="iotty",
+        )
+
+    @property
+    def current_cover_position(self) -> int | None:
+        """Return the current position of the shutter.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._iotty_device.percentage
+
+    @property
+    def is_closed(self) -> bool:
+        """Return true if the Shutter is closed."""
+        _LOGGER.debug(
+            "Retrieve device status for %s ? %s : %s",
+            self._iotty_device.device_id,
+            self._iotty_device.status,
+            self._iotty_device.percentage,
+        )
+        return (
+            self._iotty_device.status == ShutterState.STATIONARY
+            and self._iotty_device.percentage == 0
+        )
+
+    @property
+    def is_opening(self) -> bool:
+        """Return true if the Shutter is opening."""
+        return self._iotty_device.status == ShutterState.OPENING
+
+    @property
+    def is_closing(self) -> bool:
+        """Return true if the Shutter is closing."""
+        return self._iotty_device.status == ShutterState.CLOSING
+
+    @property
+    def supported_features(self) -> CoverEntityFeature:
+        """Flag supported features."""
+        return CoverEntityFeature(0) | (
+            CoverEntityFeature.OPEN
+            | CoverEntityFeature.CLOSE
+            | CoverEntityFeature.STOP
+            | CoverEntityFeature.SET_POSITION
+        )
+
+    async def async_open_cover(self, **kwargs: Any) -> None:
+        """Open the cover."""
+        await self._iotty_cloud.command(
+            self._iotty_device.device_id, self._iotty_device.cmd_open()
+        )
+        await self.coordinator.async_request_refresh()
+
+    async def async_close_cover(self, **kwargs: Any) -> None:
+        """Close cover."""
+        await self._iotty_cloud.command(
+            self._iotty_device.device_id, self._iotty_device.cmd_close()
+        )
+        await self.coordinator.async_request_refresh()
+
+    async def async_set_cover_position(self, **kwargs: Any) -> None:
+        """Move the cover to a specific position."""
+        percentage = kwargs[ATTR_POSITION]
+        await self._iotty_cloud.command(
+            self._iotty_device.device_id,
+            self._iotty_device.cmd_move_to(),
+            {"open_percentage": percentage},
+        )
+        await self.coordinator.async_request_refresh()
+
+    async def async_stop_cover(self, **kwargs: Any) -> None:
+        """Stop the cover."""
+        await self._iotty_cloud.command(
+            self._iotty_device.device_id, self._iotty_device.cmd_stop()
+        )
+        await self.coordinator.async_request_refresh()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+
+        device: Device = next(
+            device
+            for device in self.coordinator.data.devices
+            if device.device_id == self._iotty_device.device_id
+        )
```

Comment:
Would it make sense to already transform the coordinator data to a key-value dict of all devices? It will remove the need for all these lookups

Reply:
The API returns a list of devices, so every time it gets called we'd need to map it into a key-value dict, which would move a bit of the computation there, but definitely improve the performance.
Since this is a change that would impact also other parts of the code, I think it would be best suited for a later optimization PR after this one, to stay aligned with the "1 change per PR" rule